Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -56,25 +65,167 @@ namespace WebCore {
-    cookieStore->suspendIfNeeded();
-    return cookieStore;
 }
 
-CookieStore::CookieStore(Document* document)
-    : ActiveDOMObject(document)
+void CookieStore::MainThreadBridge::ensureOnMainThread(Function<void(ScriptExecutionContext&)>&& task)
 {
-    if (!document)
+    ASSERT(m_cookieStore);
+    if (!m_cookieStore)
         return;
 
-    auto* page = document->page();
-    if (!page)
+    RefPtr context = m_cookieStore->scriptExecutionContext();
+    if (!context)
+        return;
+    ASSERT(context->isContextThread());
+
+    if (RefPtr document = dynamicDowncast<Document>(*context)) {
+        task(*context);
         return;
+    }
 
-    m_cookieJar = page->cookieJar();
-    m_host = document->url().host().toString();
+    downcast<WorkerGlobalScope>(*context).thread().workerLoaderProxy().postTaskToLoader(WTFMove(task));
+}
+
+void CookieStore::MainThreadBridge::ensureOnContextThread(Function<void(CookieStore&)>&& task)
+{
+    ScriptExecutionContext::ensureOnContextThread(m_contextIdentifier, [protectedThis = Ref { *this }, task = WTFMove(task)](auto&) {
+        if (RefPtr cookieStore = protectedThis->m_cookieStore.get())
+            task(*cookieStore);
+    });
+}
+
+void CookieStore::MainThreadBridge::get(CookieStoreGetOptions&& options, Function<void(CookieStore&, ExceptionOr<Vector<Cookie>>&&)>&& completionHandler)
+{
+    ASSERT(m_cookieStore);
+    if (!m_cookieStore)
+        return;
+
+    auto getCookies = [this, protectedThis = Ref { *this }, options = crossThreadCopy(WTFMove(options)), completionHandler = WTFMove(completionHandler)](ScriptExecutionContext& context) mutable {
+        Ref document = downcast<Document>(context);
+        WeakPtr page = document->page();
+        if (!page) {
+            ensureOnContextThread([completionHandler = WTFMove(completionHandler)](CookieStore& cookieStore) mutable {
+                completionHandler(cookieStore, Exception { SecurityError });
+            });
+            return;
+        }
+
+        auto& cookieJar = page->cookieJar();
+        auto resultHandler = [this, completionHandler = WTFMove(completionHandler)] (std::optional<Vector<Cookie>>&& cookies) mutable {
+            ensureOnContextThread([completionHandler = WTFMove(completionHandler), cookies = crossThreadCopy(WTFMove(cookies))](CookieStore& cookieStore) mutable {
+                if (!cookies)
+                    completionHandler(cookieStore, Exception { TypeError });
+                else
+                    completionHandler(cookieStore, WTFMove(*cookies));
+            });
+        };
+
+        cookieJar.getCookiesAsync(document, document->url(), options, WTFMove(resultHandler));
+    };
+
+    ensureOnMainThread(WTFMove(getCookies));
+}
+
+void CookieStore::MainThreadBridge::getAll(CookieStoreGetOptions&& options, URL&& url, Function<void(CookieStore&, ExceptionOr<Vector<Cookie>>&&)>&& completionHandler)
+{
+    ASSERT(m_cookieStore);
+    if (!m_cookieStore)
```

Comment:
Can we drop these if checks in all 3 functions? The assertion should suffice given that they are called by the cookie store.